app/Http/Controllers/OAuthController.php

```diff
@@ -2,20 +2,78 @@
 
 namespace App\Http\Controllers;
 
-use App\Models\User;
-use CDash\Middleware\OAuth2\OAuth2Interface;
 use Illuminate\Http\RedirectResponse;
 use Illuminate\Http\Request;
 use Illuminate\Support\Collection;
 use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Hash;
+use Illuminate\Support\Facades\Log;
+use Illuminate\Support\Str;
+use Laravel\Socialite\Facades\Socialite;
+use Symfony\Component\HttpFoundation\RedirectResponse as symfonyResponse;
 use RuntimeException;
 
+use App\Models\User;
+use App\Http\Controllers\Auth\RegisterController;
+use CDash\Middleware\OAuth2\OAuth2Interface;
+
 /**
  * Class OAuthController
  * @package App\Http\Controllers
  */
 final class OAuthController extends AbstractController
 {
+    /**
+     * @param Request $request
+     * @param string $service
+     * @return symfonyResponse
+     */
+
+    public function socialite(Request $request, string $service)
+    {
+        return Socialite::driver($service)->redirect();
+    }
+
+    /**
+     * @param string $service
+     * @param User $user_table
+     * @return RedirectResponse
+    */
+
+    public function callback(string $service, User $user_table)
+    {
+        $authUser = Socialite::driver($service)->user();
+        $email =  $authUser->getEmail();
+        [$fname, $lname] = explode(" ", $authUser->getName());
+
+        // TODO: What if, for whatever reason, there is more than one user found?
+        $user = $user_table->whereIn('email', [$email])->first();
+        if (!$user) {
+            if(config("services.{$service}.autoregister") === true) {
+                $registerData = [
+                    'email' => $email,
+                    'fname' => $fname,
+                    'lname' => $lname,
+                    'institution' => '',
+                    'password' => Hash::make(Str::random(40)),
+                ];
+                $registerController = new RegisterController();
+                $user = $registerController->create($registerData);
+                if ($user === null) {
+                    Log::error("Error registering new SAML2 user: $email");
+                    abort(500, "Error registering new SAML2 user");
+                }
+            } else {
+                $parameters = compact('fname', "lname", 'email');
+                $to = route('register', $parameters);
+                return redirect($to);
+            }
+        } else {
+            Auth::login($user, true);
+        }
+        return redirect('/index.php');
```

Comment:
It would be better to redirect to `session('url.intended') ?? '/'` so the user gets redirected to the page they were trying to access, if applicable.